### DISCUSSION TITLE
Quadrature over part of a mesh

### DISCUSSION THREAD

**Discussion by ethrelfall** [firedrakeproject/firedrake#3365] [Firedrake support] [answered]
Dear community,

I've got a 2D domain on which I have a scalar function.  I know how to integrate the function over the entire domain in Firedrake, but it is possible to integrate it over the entire range of y but a particular range in x (e.g. from x0 to x1), please?

I don't think I want to do it by multiplying the scalar function by a window function BTW.

Search for matching discussions done.

Thanks!

**Comment by colinjcotter** [comment]
For that you need to generate a mesh (eg using gmsh or
netgen) so that the subdomain you want to integrate over is tagged. Then you provide the subdomain tag to the dx measure: dx(<tag>)

On 28 Jan 2024, at 20:44, ethrelfall ***@***.***> wrote:

﻿

Dear community,

I've got a 2D domain on which I have a scalar function. I know how to integrate the function over the entire domain in Firedrake, but it is possible to integrate it over the entire range of y but a particular range in x (e.g. from x0 to x1), please?

I don't think I want to do it by multiplying the scalar function by a window function BTW.

Search for matching discussions done.

Thanks!

**Reply by ethrelfall** [reply]
Many thanks.  I don't think the suggested will work for what I'm trying to do (kinetics, viz. 

![image](https://github.com/firedrakeproject/firedrake/assets/77395015/32e8e84e-72ff-4a46-9433-4e118792e130)

wh. f and phi are scalars in a 2D domain coordinatized by (x,v)).

In the same vein, I assume I can't do a line integral (e.g. along a vertical line in a 2D domain) without meshing the line and identifying it as a domain.

So OK, I think I can't do the above problem in FD without using fudges or being capable enough to add new features (which I'm not).

Cheers,

Ed.

**Comment by colinjcotter** [comment, marked as answer]
Ah, you are talking about the projection in Vlasov. I have a few approaches to that, probably best if we meet for a chat instead of Github discussions as it depends on discretisation choices. I’m running an MSc project on kinetics with Firedrake currently.

Re: the line. This can be discretised as a quadrature sum, and implemented using vertex only mesh. But it might be better for us to discuss what you want to do.

On 29 Jan 2024, at 08:00, ethrelfall ***@***.***> wrote:

﻿

Many thanks. I don't think the suggested will work for what I'm trying to do (kinetics, viz.

image.png (view on web)<https://github.com/firedrakeproject/firedrake/assets/77395015/32e8e84e-72ff-4a46-9433-4e118792e130>

wh. f and phi are scalars in a 2D domain coordinatized by (x,v)).

In the same vein, I assume I can't do a line integral (e.g. along a vertical line in a 2D domain) without meshing the line and identifying it as a domain.

So OK, I think I can't do the above problem in FD without using fudges or being capable enough to add new features (which I'm not).

Cheers,

Ed.

**Reply by ethrelfall** [reply]
Many thanks, maybe we can talk after meeting later this week.

**Reply by dham** [reply]
Just to add, using cross-mesh interpolation you can actually mesh the line independently of the domain and then interpolate the expression onto the line and integrate. This is hot-off-the press Firedrake functionality that has been fully merged in the last month.

**Reply by ethrelfall** [reply]
Thanks David, that sounds like just what's needed for this sort of thing.

**Reply by colinjcotter** [reply]
I actually realised after our chat today that this less complicated than I was saying. If you have an extruded mesh, then make Phi in a function space that is independent of the depth (extruded direction) using `vfamily = "R"`. Then the Phi equation (with test function from the same space) will restrict to depth independent functions.

The equation becomes

$$ \int p_x \Phi_x \,d x = \int p \omega_P^2 (H\frac{f}{v_N} - 1) \,dx, \quad \forall p \in \bar{V},$$

where $\bar{V}$ is the depth independent space, and $H$ is the height of your domain (that you are approximating for an infinite height), and 
where here the $dx$ integral means integration over space and velocity coordinates (so integrating over your whole x-v mesh as usual).

This works because everything except for $f$ is independent of depth, so you can take the vertical part of the double integral and put it around $f$ only.

**Reply by colinjcotter** [reply]
This only works for a 1D velocity coordinate though.

**Comment by ethrelfall** [comment]
Thanks Colin - I've implemented and it looks right ...
![image](https://github.com/firedrakeproject/firedrake/assets/77395015/5b8a4f48-9284-477a-a94e-0839c97da7ba)
